File: internal/domain/protocol/codec.go
package protocol

import (
	"bytes"
	"compress/flate"
	"encoding/binary"
	"encoding/json"
	"fmt"
	"io"

	"ghostwire/internal/domain"
)

const controlCompressThreshold = 256

type DecodedMessage struct {
	Kind    MessageKind
	Flags   uint16
	Payload []byte
}

func EncodeMessage(kind MessageKind, payload []byte, flags uint16) []byte {
	if len(payload) > MaxPayloadSize {
		panic(domain.NewError(domain.ErrProtocol,
			fmt.Sprintf("payload of %d bytes exceeds limit", len(payload))))
	}
	header := make([]byte, HeaderSize)
	binary.BigEndian.PutUint32(header[0:4], ProtocolMagic)
	header[4] = ProtocolVersion
	header[5] = byte(kind)
	binary.BigEndian.PutUint16(header[6:8], flags)
	binary.BigEndian.PutUint32(header[8:12], uint32(len(payload)))
	out := make([]byte, HeaderSize+len(payload))
	copy(out, header)
	copy(out[HeaderSize:], payload)
	return out
}

func EncodeJSONMessage(kind MessageKind, v any) []byte {
	var buf bytes.Buffer
	enc := json.NewEncoder(&buf)
	enc.SetEscapeHTML(false)
	if err := enc.Encode(v); err != nil {
		panic(domain.WrapError(err, domain.ErrInternal, "failed to encode JSON"))
	}
	jsonBytes := buf.Bytes()

	if len(jsonBytes) > controlCompressThreshold {
		var compressed bytes.Buffer
		w, _ := flate.NewWriter(&compressed, 1)
		w.Write(jsonBytes)
		w.Close()
		if compressed.Len() < len(jsonBytes) {
			return EncodeMessage(kind, compressed.Bytes(), uint16(FlagDeflate))
		}
	}
	return EncodeMessage(kind, jsonBytes, 0)
}

func DecodeControlPayload(msg DecodedMessage) ([]byte, error) {
	if msg.Flags&uint16(FlagDeflate) == 0 {
		return msg.Payload, nil
	}
	r := flate.NewReader(bytes.NewReader(msg.Payload))
	defer r.Close()
	out, err := io.ReadAll(r)
	if err != nil {
		return nil, domain.WrapError(err, domain.ErrProtocol, "failed to inflate control payload")
	}
	if len(out) > 1<<20 {
		return nil, domain.NewError(domain.ErrProtocol, "inflated payload too large")
	}
	return out, nil
}

type MessageDecoder struct {
	chunks   [][]byte
	head     int
	buffered int
}

func (d *MessageDecoder) Feed(chunk []byte) ([]DecodedMessage, error) {
	if len(chunk) > 0 {
		d.chunks = append(d.chunks, chunk)
		d.buffered += len(chunk)
	}

	var messages []DecodedMessage
	for {
		if d.buffered < HeaderSize {
			break
		}
		header := d.peek(HeaderSize)
		magic := binary.BigEndian.Uint32(header[0:4])
		if magic != ProtocolMagic {
			d.reset()
			return nil, domain.NewError(domain.ErrProtocol, "stream is out of sync (bad magic)")
		}
		version := header[4]
		if version != ProtocolVersion {
			d.reset()
			return nil, domain.NewError(domain.ErrVersion,
				fmt.Sprintf("unsupported protocol version %d", version))
		}
		kind := MessageKind(header[5])
		flags := binary.BigEndian.Uint16(header[6:8])
		length := binary.BigEndian.Uint32(header[8:12])

		if int(length) > MaxPayloadSize {
			d.reset()
			return nil, domain.NewError(domain.ErrProtocol,
				fmt.Sprintf("payload length %d exceeds limit", length))
		}
		if d.buffered < HeaderSize+int(length) {
			break
		}
		d.skip(HeaderSize)
		payload := d.take(int(length))
		messages = append(messages, DecodedMessage{Kind: kind, Flags: flags, Payload: payload})
	}
	return messages, nil
}

func (d *MessageDecoder) Reset() {
	d.reset()
}

func (d *MessageDecoder) reset() {
	d.chunks = d.chunks[:0]
	d.head = 0
	d.buffered = 0
}

func (d *MessageDecoder) peek(count int) []byte {
	first := d.chunks[0]
	if len(first)-d.head >= count {
		return first[d.head : d.head+count]
	}
	var parts [][]byte
	need := count
	index := 0
	for need > 0 {
		chunk := d.chunks[index]
		start := 0
		if index == 0 {
			start = d.head
		}
		available := len(chunk) - start
		take := need
		if take > available {
			take = available
		}
		parts = append(parts, chunk[start:start+take])
		need -= take
		index++
	}
	return bytes.Join(parts, nil)
}

func (d *MessageDecoder) skip(count int) {
	d.head += count
	d.buffered -= count
	d.compact()
}

func (d *MessageDecoder) take(count int) []byte {
	first := d.chunks[0]
	if len(first)-d.head >= count {
		out := first[d.head : d.head+count]
		d.head += count
		d.buffered -= count
		d.compact()
		return out
	}
	var parts [][]byte
	need := count
	for need > 0 {
		chunk := d.chunks[0]
		available := len(chunk) - d.head
		take := need
		if take > available {
			take = available
		}
		parts = append(parts, chunk[d.head:d.head+take])
		need -= take
		d.buffered -= take
		d.head += take
		if d.head == len(chunk) {
			d.chunks = d.chunks[1:]
			d.head = 0
		}
	}
	d.compact()
	return bytes.Join(parts, nil)
}

func (d *MessageDecoder) compact() {
	for len(d.chunks) > 0 && d.head >= len(d.chunks[0]) {
		d.head -= len(d.chunks[0])
		d.chunks = d.chunks[1:]
	}
}

File: src/core/domain/protocol/codec.ts
import { deflateSync, inflateSync } from 'node:zlib';
import { GhostwireError } from '../errors.js';
import {
  HEADER_FLAG,
  HEADER_SIZE,
  MAX_PAYLOAD_SIZE,
  PROTOCOL_MAGIC,
  PROTOCOL_VERSION,
} from './constants.js';

export interface DecodedMessage {
  kind: number;
  flags: number;
  payload: Uint8Array;
}

const CONTROL_COMPRESS_THRESHOLD = 256;

export function encodeMessage(kind: number, payload: Uint8Array, flags = 0): Buffer {
  if (payload.length > MAX_PAYLOAD_SIZE) {
    throw new GhostwireError('protocol', `payload of ${payload.length} bytes exceeds limit`);
  }
  const header = Buffer.allocUnsafe(HEADER_SIZE);
  header.writeUInt32BE(PROTOCOL_MAGIC, 0);
  header.writeUInt8(PROTOCOL_VERSION, 4);
  header.writeUInt8(kind, 5);
  header.writeUInt16BE(flags, 6);
  header.writeUInt32BE(payload.length, 8);
  const body = Buffer.from(payload.buffer, payload.byteOffset, payload.length);
  return Buffer.concat([header, body]);
}

export function encodeJSONMessage(kind: number, value: unknown): Buffer {
  const json = Buffer.from(JSON.stringify(value), 'utf8');
  if (json.length > CONTROL_COMPRESS_THRESHOLD) {
    const compressed = deflateSync(json, { level: 3 });
    if (compressed.length < json.length) {
      return encodeMessage(kind, compressed, HEADER_FLAG.Deflate);
    }
  }
  return encodeMessage(kind, json);
}

export function decodeControlPayload(message: DecodedMessage): Uint8Array {
  if ((message.flags & HEADER_FLAG.Deflate) === 0) return message.payload;
  try {
    const out = inflateSync(Buffer.from(message.payload.buffer, message.payload.byteOffset, message.payload.byteLength), {
      maxOutputLength: 1 << 20,
    });
    return new Uint8Array(out.buffer, out.byteOffset, out.byteLength);
  } catch (err) {
    throw new GhostwireError('protocol', 'failed to inflate control payload', { cause: err });
  }
}

export class MessageDecoder {
  private chunks: Buffer[] = [];
  private head = 0;
  private buffered = 0;

  feed(chunk: Uint8Array): DecodedMessage[] {
    const buffer =
      chunk instanceof Buffer
        ? chunk
        : Buffer.from(chunk.buffer, chunk.byteOffset, chunk.byteLength);
    if (buffer.length > 0) {
      this.chunks.push(buffer);
      this.buffered += buffer.length;
    }
    const messages: DecodedMessage[] = [];
    for (;;) {
      if (this.buffered < HEADER_SIZE) break;
      const header = this.peek(HEADER_SIZE);
      if (header.readUInt32BE(0) !== PROTOCOL_MAGIC) {
        this.reset();
        throw new GhostwireError('protocol', 'stream is out of sync (bad magic)');
      }
      const version = header.readUInt8(4);
      if (version !== PROTOCOL_VERSION) {
        this.reset();
        throw new GhostwireError('version', `unsupported protocol version ${version}`);
      }
      const kind = header.readUInt8(5);
      const flags = header.readUInt16BE(6);
      const length = header.readUInt32BE(8);
      if (length > MAX_PAYLOAD_SIZE) {
        this.reset();
        throw new GhostwireError('protocol', `payload length ${length} exceeds limit`);
      }
      if (this.buffered < HEADER_SIZE + length) break;
      this.skip(HEADER_SIZE);
      const payload = this.take(length);
      messages.push({ kind, flags, payload });
    }
    return messages;
  }

  reset(): void {
    this.chunks = [];
    this.head = 0;
    this.buffered = 0;
  }

  private peek(count: number): Buffer {
    const first = this.chunks[0]!;
    if (first.length - this.head >= count) {
      return first.subarray(this.head, this.head + count);
    }
    const parts: Buffer[] = [];
    let need = count;
    let index = 0;
    while (need > 0) {
      const chunk = this.chunks[index]!;
      const start = index === 0 ? this.head : 0;
      const available = chunk.length - start;
      const take = Math.min(need, available);
      parts.push(chunk.subarray(start, start + take));
      need -= take;
      index++;
    }
    return Buffer.concat(parts);
  }

  private skip(count: number): void {
    this.head += count;
    this.buffered -= count;
    this.compact();
  }

  private take(count: number): Uint8Array {
    const first = this.chunks[0]!;
    if (first.length - this.head >= count) {
      const out = first.subarray(this.head, this.head + count);
      this.head += count;
      this.buffered -= count;
      this.compact();
      return out;
    }
    const parts: Buffer[] = [];
    let need = count;
    while (need > 0) {
      const chunk = this.chunks[0]!;
      const available = chunk.length - this.head;
      const take = Math.min(need, available);
      parts.push(chunk.subarray(this.head, this.head + take));
      need -= take;
      this.buffered -= take;
      this.head += take;
      if (this.head === chunk.length) {
        this.chunks.shift();
        this.head = 0;
      }
    }
    this.compact();
    return Buffer.concat(parts);
  }

  private compact(): void {
    while (this.chunks.length > 0 && this.head >= this.chunks[0]!.length) {
      this.chunks.shift();
      this.head = 0;
    }
  }
}
